crates/env/src/api.rs

```diff
@@ -491,24 +491,14 @@ where
 ///
 /// # Usage
 ///
-/// Normally contracts define their own `enum` dispatch types respective
-/// to their exported constructors and messages that implement `scale::Decode`
-/// according to the constructors or messages selectors and their arguments.
-/// These `enum` dispatch types are then given to this procedure as the `T`.
+/// Normally contracts decoding(`scale::Decode`) their arguments according to the
+/// constructors or messages.
 ///
-/// When using ink! users do not have to construct those enum dispatch types
-/// themselves as they are normally generated by the ink! code generation
-/// automatically.
-///
-/// # Errors
-///
-/// If the given `T` cannot be properly decoded from the expected input.
-pub fn decode_input<T>() -> Result<T>
-where
-    T: scale::Decode,
-{
-    <EnvInstance as OnInstance>::on_instance(|instance| {
-        EnvBackend::decode_input::<T>(instance)
+/// When using ink! users do not have to decode input themselves as they are
+/// normally decoded by the ink! code generation automatically.
+pub fn input_bytes<'a>() -> &'a [u8] {
+    <EnvInstance as OnInstance<'a>>::on_instance(|instance: &'a mut EnvInstance| {
+        EnvBackend::input_bytes(instance)
     })
```

Comment:
This API as stated is extremely unsafe to use by design. This is exactly why the `decode_input` API exists as is. Just notice how with this you could safely (!!) do the following:
```rust
let input = ink_env::input_bytes(); // Returns a slice refering to ink!'s static buffer
let caller = ink_env::caller(); // Mutates ink!'s static buffer to hold an AccountId
// at this point the above input binding no longer holds the input but the AccountId
```
I think I do not have to point out how extremely unsafe this API design is and therefore I would never accept a PR introducing it, even if it was marked unsafe since we do not want to introduce unsafe APIs if we can do better.

Reply:
The idea was that borrow check will prevent you from the usage of `ink_env::caller()` if you will use `input` after that. But I tested it now and... It allows that=D So yes, it should be changed.

Do you have idea how better to provide the usage of static buffer to the user in safety way?

Reply:
> Do you have idea how better to provide the usage of static buffer to the user in safety way?

Well I designed the `decode_input` API exactly this way because of exactly this issue. So my proposal is to keep `decode_input` API as is.

Reply:
But the idea of the change is to decode directly in the place where we want to use arguments. Instead of decoding them in another place and then moving them to the place where we want to use them=)

Reply:
I think a better idea is to still use `decode_input` but not move the arguments and instead use references where the user wants to use references. Let me refer to [this issue](https://github.com/paritytech/ink/issues/989) for this use case.

Reply:
What I experienced while rewriting the entire codegen is that the differentiation between decoding the messages and executing the messages caused a minor Wasm size bloat that could probably be fixed without changing the `decode_input` API at all. I haven't tried yet myself though.

Reply:
Concerning `return_bytes` I'd also like to keep the previous `return_value` with some built-in optimizations that can be used to avoid ink!'s static buffer. We could use `parity_scale_codec::Encode::encoded_fixed_size` method for that for example.

Reply:
> I think a better idea is to still use decode_input but not move the arguments and instead use references where the user wants to use references. Let me refer to this issue for this use case.


Apart from optimizing the size, I am also pursuing another idea: Decode input arguments directly in the message to be agnostic to other messages. At the moment you need to know the list of all messages from all traits and inherent implementation to create decodable enum.

This restriction doesn't allow to support default implementation in traits, because users should enumerate all methods in the implementation section(it breaks the idea of default implementation).

```rust
// This line allows the user to inherit default implementation of the trait
#[ink(impl)]
impl Erc20 for Ers20Struct {}
```

If methods decode the input by self. We can add a method to the trait:
```rust
pub trait Erc20: ContractEnv {
    ...
    fn execute_input(selector: [u8; 4], input: &[u8]) {
        match selector {
        ...    
        }
    }
}
```

Reply:
> Concerning return_bytes: I'd also like to keep the previous return_value with some built-in optimizations that can be used to avoid ink!'s static buffer. We could use parity_scale_codec::Encode::encoded_fixed_size method for that for example. This way we keep a high-level API and have the same performance enhancements as with your proposed low-level API.

I've already removed the usage of `return_bytes`. But with access to a static buffer(or to another buffer where we can store the encoded result), [it](https://github.com/Supercolony-net/ink/blob/d2b7a5275854d627100ae52c0d8b671e6b27c218/crates/lang/codegen/src/generator/dispatch.rs#L587) can be optimized more. It is a common part for all messages, and we can do only one `::ink_storage::traits::push_spread_root` instead of several=)

Reply:
> But with access to a static buffer(or to another buffer where we can store the encoded result), it can be optimized more.

I highly doubt this statement but cannot prove it at the moment. Using a static buffer should not at all be different performance wise from a local buffer in Wasm where we operate in a single threaded environment.

Reply:
> If methods decode the input by self. We can add a method to the trait:

Adding artificial methods to ink! trait definitions is a no-go. We must make sure that ink! modifies trait definitions (and others) as minimally as possible.

Reply:
> I highly doubt this statement but cannot prove it at the moment. Using a static buffer should not at all be different performance wise from a local buffer in Wasm where we operate in a single threaded environment.

I meant that we can change code-generation and optimize it do one `::ink_storage::traits::push_spread_root` at the end of `execute_dispatchable` instead of the end of each message section.

BTW, It should have the same performance. But in the case of a local buffer you need to allocate that memory in the stack - it is additional operations in Runtime.

Reply:
> I meant that we can change code-generation and optimize it do one ::ink_storage::traits::push_spread_root at the end of execute_dispatchable instead of the end of each message section.

Yes, there is a lot of optimization potential in the way we currently dispatch ink! constructors and messages but none of which requires changes in the `ink_env` APIs imo. That's what I meant.

Reply:
> Adding artificial methods to ink! trait definitions is a no-go. We must make sure that ink! modifies trait definitions (and others) as minimally as possible.

I agree with it. But I think that this is the only possible solution at the moment if we want to support all the features of traits in rust. Because if the user defines type alias, or constant, or super trait, you will not able to do a strick as you did for `__ink_TraitInfo`=)

Reply:
I think we should not discuss all at once. Please let us focus on the improvements to the dispatch logic in this PR.

Reply:
I mentioned trait because it requires the next logic:
First I need to get 4 bytes to find a selector. After I know which method I need to call that method will decode the remaining bytes into arguments.

So it is why I'm trying to decode them agnoticly=)

What do you think about two static buffers? One will be always read-only and will contain input bytes. Another will be used for encoding/decoding. In this case, the user will have the access to the input at any time. And he can decode what he wants by himself.

Reply:
We always get the whole input for the user(even if he wants only 4 bytes we will read all 16 MB). And if he wants to have a simple signature of method(with 0 arguments) but later he wants to decode some parts of the input, he again should get the whole data. And it can be expensive to call `ext::seal_input` each time. So I think idea with second array only for input is not bad(=